File: src/main/java/org/seminify/application/file/FileService.java
package org.seminify.application.file;

import java.io.IOException;
import java.net.MalformedURLException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardCopyOption;
import java.util.List;
import org.springframework.core.io.Resource;
import org.springframework.core.io.UrlResource;
import org.springframework.stereotype.Service;
import org.springframework.util.FileSystemUtils;
import org.springframework.web.multipart.MultipartFile;

@Service
public class FileService {

  private final Path path;

  public FileService(FileProperties fileProperties) {
    if (
      fileProperties.getPath().trim().length() == 0
    ) throw new RuntimeException("File upload location can not be Empty.");
    this.path = Paths.get(fileProperties.getPath());
  }

  public List<Path> paths() {
    try {
      return Files.walk(path, 1)
        .filter(arg0 -> !arg0.equals(path))
        .map(path::relativize)
        .toList();
    } catch (IOException e) {
      throw new RuntimeException("Failed to read stored files", e);
    }
  }

  public Resource resource(String filename) {
    try {
      var resource = new UrlResource(path.resolve(filename).toUri());
      if (resource.exists() || resource.isReadable()) return resource;
      throw new FileNotFoundException("Could not read file: " + filename);
    } catch (MalformedURLException e) {
      throw new FileNotFoundException("Could not read file: " + filename, e);
    }
  }

  public void store(MultipartFile multipartFile) {
    if (multipartFile.isEmpty()) throw new RuntimeException(
      "Failed to store empty file."
    );
    var path = this.path.resolve(Paths.get(multipartFile.getOriginalFilename()))
      .normalize()
      .toAbsolutePath();
    if (
      !path.getParent().equals(this.path.toAbsolutePath())
    ) throw new RuntimeException(
      "Cannot store file outside current directory."
    );
    try (var inputStream = multipartFile.getInputStream()) {
      Files.copy(inputStream, path, StandardCopyOption.REPLACE_EXISTING);
    } catch (Exception e) {
      throw new RuntimeException("Failed to store file.", e);
    }
  }

  public void deleteAll() {
    FileSystemUtils.deleteRecursively(path.toFile());
  }

  public void init() {
    try {
      Files.createDirectories(path);
    } catch (IOException e) {
      throw new RuntimeException("Could not initialize storage", e);
    }
  }
}

File: src/main/kotlin/org/seminify/application/file/FileService.kt
package org.seminify.application.file

import org.springframework.core.io.Resource
import org.springframework.core.io.UrlResource
import org.springframework.stereotype.Service
import org.springframework.util.FileSystemUtils
import org.springframework.web.multipart.MultipartFile
import java.io.FileNotFoundException
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.nio.file.StandardCopyOption

@Service
class FileService(fileProperties: FileProperties) {
    private val path: Path

    init {
        if (fileProperties.path.trim().isEmpty())
            throw RuntimeException("File upload location can not be Empty.")
        path = Paths.get(fileProperties.path)
    }

    fun paths(): List<Path> {
        return Files.walk(path, 1).filter { it != path }.map { it.relativize(path) }.toList()
    }

    fun resource(filename: String): Resource {
        val resource = UrlResource(this.path.resolve(filename).toUri())
        if (resource.exists() || resource.isReadable)
            return resource
        else throw FileNotFoundException("Could not read file: $filename")
    }

    fun store(multipartFile: MultipartFile) {
        if (multipartFile.isEmpty)
            throw RuntimeException("Failed to store empty file.")
        val path = this.path.resolve(
            multipartFile.originalFilename ?: throw RuntimeException("Failed to store file with null filename")
        ).normalize().toAbsolutePath()
        if (!path.parent.equals(this.path.toAbsolutePath()))
            throw RuntimeException("Cannot store file outside current directory.")
        multipartFile.inputStream.use {
            Files.copy(it, path, StandardCopyOption.REPLACE_EXISTING)
        }
    }

    fun deleteAll() {
        FileSystemUtils.deleteRecursively(path.toFile())
    }

    fun init() {
        Files.createDirectories(path)
    }
}
